components/configable/config_derive/src/lib.rs

```diff
@@ -0,0 +1,156 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! This crate provides a macro that can be used to generate code to
+//! implement `Configable` trait
+//!
+
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::punctuated::Punctuated;
+use syn::token::Comma;
+use syn::*;
+
+#[proc_macro_derive(Configable, attributes(config))]
+pub fn config(input: TokenStream) -> TokenStream {
+    let ast = parse_macro_input!(input as DeriveInput);
+    match generate_token(ast.ident, ast.data) {
+        Ok(res) => res.into(),
+        Err(e) => TokenStream::from(e.to_compile_error()),
+    }
+}
+
+fn generate_token(name: Ident, data: Data) -> std::result::Result<proc_macro2::TokenStream, Error> {
+    let creat_name = Ident::new("configable", proc_macro2::Span::call_site());
+    let fields = get_struct_fields(name.span(), &data)?;
+    let update_fn = update(&fields, &creat_name)?;
+    let diff_fn = diff(&fields, &creat_name)?;
+    Ok(quote! {
+        impl #creat_name::Configable for #name {
+            #update_fn
+            #diff_fn
+        }
+    })
+}
+
+fn get_struct_fields(
+    span: proc_macro2::Span,
+    data: &Data,
+) -> std::result::Result<&Punctuated<Field, Comma>, Error> {
+    if let Data::Struct(DataStruct {
+        fields: Fields::Named(FieldsNamed { ref named, .. }),
+        ..
+    }) = data
+    {
+        Ok(named)
+    } else {
+        Err(Error::new(
+            span,
+            "expect derive Configable on struct with named fields!",
+        ))
+    }
+}
+
+fn update(
+    fields: &Punctuated<Field, Comma>,
+    creat_name: &Ident,
+) -> Result<proc_macro2::TokenStream> {
+    let incoming = Ident::new("incoming", proc_macro2::Span::call_site());
+    let mut update_fields = Vec::with_capacity(fields.len());
+    for field in fields {
+        let (not_support, submodule) = get_attrs(&field.attrs)?;
+        if not_support || field.ident.is_none() {
+            continue;
+        }
+        let name = field.ident.as_ref().unwrap();
+        let name_lit = LitStr::new(&format!("{}", name), name.span());
+        let f = if submodule {
+            quote! {
+                if let Some(#creat_name::ConfigValue::Module(v)) = #incoming.remove(#name_lit) {
+                    #creat_name::Configable::update(&mut self.#name, v);
+                }
+            }
+        } else {
+            quote! {
+                if let Some(v) = #incoming.remove(#name_lit) {
+                    self.#name = v.into();
+                }
+            }
+        };
+        update_fields.push(f);
+    }
```

Comment:
Instead of expanding code at compile-time, could you fold them into a static array and iterate them at run-time? This helps reducing compiling time and binary size.